blob/gcsblob/iam.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This is an implementation for Options.MakeSignBytes
+// that serves as example for how to keep a private key in a separate
+// process, service, or HSM/TPM, yet use it as signer for blob.Bucket.
+
+package gcsblob
+
+import (
+	"context"
+	"sync"
+
+	credentials "cloud.google.com/go/iam/credentials/apiv1"
+	gax "github.com/googleapis/gax-go/v2"
+	credentialspb "google.golang.org/genproto/googleapis/iam/credentials/v1"
+)
+
+// credentialsClient wraps the IAM Credentials API client for a lazy initialization
+// and expresses it in the reduced format expected by SignBytes.
+// See https://cloud.google.com/iam/docs/reference/credentials/rest
+type credentialsClient struct {
```

Comment:
I am fine with this for now, but when we make it public I don't think it should be a struct like this, it should just be a function as I suggested in my previous comment.

* Creating a new client is not expensive, I don't think it even blocks.
* The lazy URLOpener is only called if someone is instantiating a URL using it; they can't override the defaults after that.
* The lazy URLOpener already has code supporting the `sync.Once` stuff, it's unnecessary to implement it again here.